oximeter/db/src/client.rs

```diff
@@ -267,14 +271,301 @@ impl Client {
+            }
+            current = *version;
+            self.insert_version(current).await?;
+        }
+        Ok(())
+    }
+
+    fn verify_schema_upgrades(
+        files: &BTreeMap<String, (PathBuf, String)>,
+    ) -> Result<(), Error> {
+        let re =
+            regex::Regex::new("(INSERT INTO)|(ALTER TABLE .* DELETE)").unwrap();
+        for (path, sql) in files.values() {
+            if re.is_match(&sql) {
+                return Err(Error::SchemaUpdateModifiesData {
+                    path: path.clone(),
+                    statement: sql.clone(),
+                });
+            }
+            if sql.matches(';').count() > 1 {
+                return Err(Error::MultipleSqlStatementsInSchemaUpdate {
+                    path: path.clone(),
+                });
+            }
+        }
+        Ok(())
+    }
+
+    async fn apply_one_schema_upgrade(
+        &self,
+        replicated: bool,
+        next_version: u64,
+        schema_dir: impl AsRef<Path>,
+    ) -> Result<(), Error> {
+        let schema_dir = schema_dir.as_ref();
+        let upgrade_file_contents = Self::read_schema_upgrade_sql_files(
+            &self.log,
+            replicated,
+            next_version,
+            schema_dir,
+        )
+        .await?;
+
+        // We need to be pretty careful at this point with any data-modifying
+        // statements. There should be no INSERT queries, for example, which we
+        // check here. ClickHouse doesn't support much in the way of data
+        // modification, which makes this pretty easy.
+        Self::verify_schema_upgrades(&upgrade_file_contents)?;
+
+        // Apply each file in sequence in the upgrade directory.
+        for (name, (path, sql)) in upgrade_file_contents.into_iter() {
+            debug!(
+                self.log,
+                "apply schema upgrade file";
+                "version" => next_version,
+                "path" => path.display(),
+                "filename" => &name,
+            );
+            match self.execute(sql).await {
+                Ok(_) => debug!(
+                    self.log,
+                    "successfully applied schema upgrade file";
+                    "version" => next_version,
+                    "path" => path.display(),
+                    "name" => name,
+                ),
+                Err(e) => {
+                    return Err(e);
+                }
+            }
+        }
+        Ok(())
+    }
+
+    // Read all SQL files, in order, in the schema directory for the provided
+    // version.
+    async fn read_schema_upgrade_sql_files(
+        log: &Logger,
+        replicated: bool,
+        version: u64,
+        schema_dir: impl AsRef<Path>,
+    ) -> Result<BTreeMap<String, (PathBuf, String)>, Error> {
+        let version_schema_dir = schema_dir
+            .as_ref()
+            .join(if replicated { "replicated" } else { "single-node" })
+            .join(version.to_string());
```

Comment:
Let's make a helper function out of this, since it's being repeated in a function above